### ISSUE TITLE
LinearLR Scheduler returning incorrect learning rate

### ISSUE THREAD

**Issue by Toukenize** [pytorch/pytorch#67426]
## 🐛 Bug

LinearLR scheduler is not returning the correct learning rate. See current vs expected learning rate schedule plots below for more details.

## To Reproduce

1. Initialize an optimizer (any), set its learning rate to something easy to debug (e.g. 1.0)
2. Initialize scheduler using `LinearLR`, set `total_iters` to something small (e.g. 5)
3. Step through the scheduler, record the learning rates at each step & visualize it
 
Code to reproduce:

```
import torch.nn as nn
from torch.optim import AdamW
from torch.optim.lr_scheduler import LinearLR

import matplotlib.pyplot as plt

# Init dummy optimizer
lr = 1
lin = nn.Linear(10,10)
opt = AdamW(lin.parameters(), lr=lr)

# Store scheduler learing rates for different `total_iters`
start_factor = 0.1
end_factor = 1.
lr_dict = dict()

for ti in [3, 5, 10, 25, 50, 100]:
    sch = LinearLR(
        opt,
        start_factor=start_factor,
        end_factor=end_factor,
        total_iters=ti,
        last_epoch=-1
    )
    
    lr_dict[ti] = []
    
    for i in range(150):
        lr_dict[ti].append(sch.get_lr())
        sch.step()

# Visualize plot
plt.figure(facecolor='white')

for total_iter, lrs in lr_dict.items():
    plt.plot(lrs, label=total_iter)
    
plt.xlabel('Steps')
plt.ylabel('Learning Rate')
plt.title(f'Current Behaviour - LinearLR Scheduler (Init LR = {lr})')
plt.legend(
    title='Total Iters',
    facecolor='white',
    bbox_to_anchor=[1.01, 0.5],
    loc='center left'
)
plt.show()
```
![image](https://user-images.githubusercontent.com/43180977/139194238-ff1562c8-86fd-4a13-90fd-f7c3f35104db.png)

## Expected behavior

Learning rate should ramp up linearly till `total_iters`

![image](https://user-images.githubusercontent.com/43180977/139195112-9464a84d-4913-4dd5-a421-f0f414978611.png)

## Environment

Please copy and paste the output from our
[environment collection script](https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py)
(or fill out the checklist below manually).

```
PyTorch version: 1.10.0+cu102
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.3 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.10

Python version: 3.7.6 (default, Jan  8 2020, 19:59:22)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-5.11.0-38-generic-x86_64-with-debian-bullseye-sid
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: GPU 0: GeForce RTX 2080 Super with Max-Q Design
Nvidia driver version: 460.91.03
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.1.1
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.1.1
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.21.2
[pip3] pytorch-lightning==1.4.9
[pip3] torch==1.10.0
[pip3] torchmetrics==0.5.1
[pip3] torchvision==0.11.1
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.0                      166  
[conda] mkl-service               2.3.0            py37he904b0f_0  
[conda] mkl_fft                   1.0.15           py37ha843d7b_0  
[conda] mkl_random                1.1.0            py37hd6b4f25_0  
[conda] numpy                     1.18.1           py37h4f9e942_0  
[conda] numpy-base                1.18.1           py37hde5b4d6_1  
[conda] numpydoc                  0.9.2                      py_0
```

You can get the script and run it with:
```
wget https://raw.githubusercontent.com/pytorch/pytorch/master/torch/utils/collect_env.py
# For security purposes, please check the contents of collect_env.py before running it.
python collect_env.py
```

 - PyTorch Version (e.g., 1.0):
 - OS (e.g., Linux):
 - How you installed PyTorch (`conda`, `pip`, source):
 - Build command you used (if compiling from source):
 - Python version:
 - CUDA/cuDNN version:
 - GPU models and configuration:
 - Any other relevant information:

## Additional context

cc @vincentqb @jbschlosser @albanD

**Comment by Toukenize** [comment]
Any updates on this?

**Comment by albanD** [comment]
HI,

Sorry for the delay. I think the problem is that you should be using `get_last_lr()` to get the last computed learning rate as per the warning that you see in your script.
The `get_lr()` is expected to be called within the step function itself only.

**Comment by Toukenize** [comment]
Hi @albanD you are right, with `get_last_lr()` everything is working as expected.

Closing issue.
